src/org/datafoodconsortium/connector/codegen/ruby/static/lib/datafoodconsortium/connector/skos_parser.rb

```diff
@@ -87,7 +90,8 @@ def parse(data)
     def createSKOSConcept(element)
         skosConcept = DataFoodConsortium::Connector::SKOSConcept.new(element.id)
         skosConcept.semanticType = element.type
-        return skosConcept
+        self.class.concepts[element.id] = skosConcept
```

Comment:
We need this change so we can correctly match Product Types and OFN taxon:
```suggestion
        id = element.id.sub(
          "https://github.com/datafoodconsortium/taxonomies/releases/latest/download/productTypes.rdf#",
          "dfc-pt:"
        )
        self.class.concepts[id] = skosConcept
```

Reply:
Okay, I will do this. But wouldn't it be better if we worked with absolute URIs in OFN as well? Less chance for any collision. People may use a different context.

Reply:
You comment prompted me to have another look, and I think I got it back to front. 
https://github.com/datafoodconsortium/taxonomies/releases/latest/download/productTypes.rdf gives us the product types with the URI, and doesn't provide any context. So we should ignore this change in the parser. 
But for the importer to work, we need to fix `skos_concept` as below, because we know `dfc-pt:` refers to productTypes. 

```ruby
      def skos_concept(object)
        return unless object.value.match(":") || object.uri?

        id = object.value.sub(
          "http://static.datafoodconsortium.org/data/measures.rdf#", "dfc-m:"
        ).sub(
          "https://github.com/datafoodconsortium/taxonomies/releases/latest/download/measures.rdf#",
          "dfc-m:"
        ).sub( # FIX HERE : 
          "dfc-pt:",
          "https://github.com/datafoodconsortium/taxonomies/releases/latest/download/productTypes.rdf#"
        )
        SKOSParser.concepts[id]
      end
```

That said, I am wondering if the importer should be aware of the given context, ie: https://github.com/openfoodfoundation/openfoodnetwork/blob/e2aeb18359eb1fadc0277bd93b0b47280cb8d8a5/engines/dfc_provider/spec/lib/data_food_consortium/connector/importer_spec.rb#L43-L60. In this example the context gives us a `static.datafoodconsortium.org` URI for `dfc-pt`, that's what we should be using for the substitution in the importer instead of the hardcoded `github.com/datafoodconsortium/taxonomies` URI. What do you think ?

Reply:
> I am wondering if the importer should be aware of the given context

Yes, and the RDF gem does that by resolving everything to absolute URIs. People could use a context in which they shorten dfc-pt just to pt or it may get renamed to something else. The context can change and just provides aliases. The URIs are the *unique resource identifiers*.

> hardcoded github.com/datafoodconsortium/taxonomies URI

The hard-coding is just a workaround for two issues:

* The URI base changed and I wanted to support the old version as well. But I don't think that's necessary anymore.
* Measures and product types were in some cases parsed with the abbreviated URI instead of the full URI. But I think that is resolved. Measures use a context and product types are all absolute.

Internally, within private methods of the connector, we can use shortened URIs based on *our* known context. For example, in the importer we can do this:
```rb
key = RDF::URI.new(uri).pname(prefixes: Context::VERSION_1_8)
```

But as soon as people start to load other vocabularies with other contexts we will have a mix of shortened DFC URIs and absolute foreign URIs. Not sure if that matters though.

The code generator used to store absolute URIs of the semantic types in the classes but now it's using the shortened version, for example:
```rb
DataFoodConsortium::Connector::Address::SEMANTIC_TYPE == "dfc-b:Address"
```

This actually means that you have to include these prefixes in the context given to the exporter. This is not a critical limitation but someone may bump into this at some point. The upside is that it's easier to read.

Anyway, this question still remains: *Do we use absolute or shortened URIs to store SKOS concepts?*

Both would work. But since all concepts are loaded from files, I consider them *external* and the RDF parser provides absolute URIs. It's less work for us to use those URIs and it's less ambiguous as well. We can still provide a lookup with prefix as well for more compatibility.

What do you think?